pkg/tracing/tracing.go

```diff
@@ -0,0 +1,181 @@
+
+	// Disable tracing immediately
+	config.SetTracingEnabled(false)
+	otel.SetTracerProvider(traceNoop.NewTracerProvider())
+
+	if shutdownErr := ce.Shutdown(ctx); shutdownErr != nil {
+		log.Errorf("Failed to shut down the tracer provider: %v", shutdownErr)
+	}
+
+	// Start a timer to re-enable tracing after one hour
+	time.AfterFunc(1*time.Hour, func() {
+		fmt.Println("Attempting to re-enable tracing...")
+		log.Info("Attempting to re-enable tracing...")
+		// Re-enable tracing
+		config.SetTracingEnabled(true)
+		// Re-initialize tracing
+		InitTracing(serviceName)
+	})
+}
+
+// Shutdown implements trace.SpanExporter.
+func (ce *customExporter) Shutdown(ctx context.Context) error {
+	log.Info("Shutting down the exporter")
+	return ce.exporter.Shutdown(ctx)
+}
+
+func getServiceNameFromSpan(span trace.ReadOnlySpan) string {
+	for _, resource := range span.Resource().Attributes() {
+		if resource.Key == "service.name" {
+			return resource.Value.AsString()
+		}
+	}
+	return config.GetTracingServiceName()
+}
+
+func (ce *customExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
+	err := ce.exporter.ExportSpans(ctx, spans)
+	ce.lock.Lock()
+	defer ce.lock.Unlock()
+
+	if err != nil {
+		ce.failureCount++
+		log.Errorf("Traces export failed: %v", err)
+
+		// Check if the failure threshold is exceeded.
+		if ce.failureCount >= 3 {
+			serviceName := getServiceNameFromSpan(spans[0])
+			ce.disableTracingTemporarily(ctx, serviceName)
+		}
+	} else {
+		// Reset failure count on a successful export.
+		ce.failureCount = 0
+	}
+
+	return err
+}
+
+// InitTracing initializes the OpenTelemetry tracing.
+func InitTracing(serviceName string) func() {
+
+	if !config.IsTracingEnabled() {
+		log.Info("Tracing is disabled")
+		return func() {}
+	}
+
+	if serviceName == "" {
+		log.Errorf("Service name was not provided in config, assuming a default service name as siglens")
+		serviceName = "siglens"
+	}
+
+	if config.GetTracingEndpoint() == "" {
+		log.Errorf("Tracing endpoint is required to initialize tracing. Disabling tracing. Please set the endpoint in the config file.")
+		config.SetTracingEnabled(false)
+		return func() {}
+	}
+
+	ctx := context.Background()
+
+	client := otlptracehttp.NewClient(
+		otlptracehttp.WithEndpointURL(config.GetTracingEndpoint()), // Your collector endpoint
+		otlptracehttp.WithInsecure(),                               // Use WithInsecure for non-production environments
+	)
+
+	exporter, err := otlptrace.New(ctx, client)
+	if err != nil {
+		log.Errorf("Failed to create the trace exporter: %v", err)
+	}
+
+	customExporter := &customExporter{
+		exporter: exporter,
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(serviceName),
+		),
+	)
+	if err != nil {
+		log.Errorf("Failed to create resource: %v", err)
+	}
+
+	tp := trace.NewTracerProvider(
+		// trace.WithSampler(trace.AlwaysSample()), // Change this to trace.NeverSample() for production
```

Comment:
I think it would be good to use https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TraceIDRatioBased and have a config variable to specify what fraction of traces should be sampled. Then people can set that config variable as appropriate for their dev, prod, etc. workload

Reply:
Got it.